modules/nextflow/src/main/groovy/nextflow/processor/TaskProcessor.groovy

```diff
@@ -803,16 +803,18 @@ class TaskProcessor {
         while( true ) {
             hash = HashBuilder.defaultHasher().putBytes(hash.asBytes()).putInt(tries).hash()
 
-            Path resumeDir = null
+            Path workDir = null
             boolean exists = false
             try {
                 final entry = session.cache.getTaskEntry(hash, this)
-                resumeDir = entry ? FileHelper.asPath(entry.trace.getWorkDir()) : null
-                if( resumeDir )
-                    exists = resumeDir.exists()
+                workDir = entry
+                    ? FileHelper.asPath(entry.trace.getWorkDir())
+                    : task.getWorkDirFor(hash)
+                if( workDir )
+                    exists = workDir.exists()
 
-                log.trace "[${safeTaskName(task)}] Cacheable folder=${resumeDir?.toUriString()} -- exists=$exists; try=$tries; shouldTryCache=$shouldTryCache; entry=$entry"
-                final cached = shouldTryCache && exists && entry.trace.isCompleted() && checkCachedOutput(task.clone(), resumeDir, hash, entry)
+                log.trace "[${safeTaskName(task)}] Cacheable folder=${workDir?.toUriString()} -- exists=$exists; try=$tries; shouldTryCache=$shouldTryCache; entry=$entry"
+                final cached = shouldTryCache && exists && entry && entry.trace.isCompleted() && checkCachedOutput(task.clone(), workDir, hash, entry)
```

Comment:
This is one of that piece of code the lesser the changes the better, both to make it simple to review and to keep history readable

Reply:
I know. This is about as simple as it gets while keeping the intent clear.

1. check for a cache entry and it's work dir, or compute the work dir if there is no cache entry
2. check whether the work dir exists
3. check for cached outputs if the cache entry and work dir exist and the task completed

and further down:
4. if the outputs are cached then use them
5. otherwise, if the work dir exists then use a new work dir
6. otherwise, create the work dir and use it

Reply:
ok, getting better, but why adding `entry &&` in the `cached` condition? If the intent is to applied the same logic when the `entry` is missing should not the condition remain the same?

Reply:
In the previous logic, it wouldn't check for cached outputs if the cache entry was missing. That is also how it behaves here. The `&& entry` is required to prevent a null reference exception on `entry.trace.isCompleted()`